### DISCUSSION TITLE
How many UEs can be created and connected?

### DISCUSSION THREAD

**Discussion by wj3451** [aligungr/UERANSIM#441] [General]
Have you tried how many UEs could be connected?

What is the limitation for the maximum UE numbers, memories or socket prot number or other resources?

Best regards

**Comment by aligungr** [comment]
Here's the theoretical limits, for maximum number of
- UEs for a single nr-ue process is **512**
- nr-ue process for a single machine is **64,512**
- UEs for a single gNodeB is **2,147,483,648**.

In other words, with `nr-ue -c myconfig.yaml -n <number>` command the limit is 512, but you can run this command as many as you want. 

---

So, in practice it's almost infinity, **but most of the time the limitation is determined by the hardware power of the machine.**

**Reply by wj3451** [reply]
Hi Aligungr

We have tried to run 3000 UEs in one Linux server only for signaling, no pdu session establishement procedure,  and it failed at about create 2000 UEs with reason socket send failed or receive failed.

We found that for each UE process, it will create two UDP sockets, one for control-plane data, and another for user-plane data, so for 2000 UEs there are about 4000 sockets running on the server, we have modified the socket number on the server but it is still failed, it could be related to other OS resources.

My question is that is there any special usage with this implementation for the coming new feature？

Can we use one UDP socket to transmit control-plane messages for all the UEs, same as user-plane data, and separate the message for each UE with unique Id like UeId+Imsi, which will save a lot of OS resource.

Thanks.

Best regards.

**Reply by aligungr** [reply]
Sure, this can be implemented as a new feature. But for now, you can go for a workaround like the following:

- Each UE node uses exactly one UDP port for radio link simulation. (Actually both control plane and user plane are carried through the same port)
- And  Each UE process uses exactly one command line port. And this is shared with the other UE nodes inside of the same process.

So, if you run 2000 UEs normally, then 4000 ports are used. However if you combine UEs in one or more processes, you save from the command line ports.

For example if you start 2000 UEs by executing `nr-ue -n 500` **4 times**, then `2000 + 4 = 2004` ports are used instead of 4000.

In this way you can cut the number of ports used to the 1/2
